### PR TITLE
automated few test cases in test_setting:

### LINKED ISSUE #5557: test_positive_update_email_reply_address
Check email reply address is updated

        :id: cb0907d1-9cb6-45c4-b2bb-e2790ea55f16

        :expectedresults: email_reply_address is updated

        :caseimportance: low

        :caseautomation: notautomated

### LINKED ISSUE #5558: test_negative_update_email_reply_address
Check email reply address is not updated

        :id: 2a2220c2-badf-47d5-ba3f-e6329930ab39

        :steps: provide invalid email addresses

        :expectedresults: email_reply_address is not updated

        :caseimportance: low

        :caseautomation: notautomated

### LINKED ISSUE #5559: test_positive_update_email_subject_prefix
Check email subject prefix is updated

        :id: c8e6b323-7b39-43d6-a9f1-5474f920bba2

        :expectedresults: email_subject_prefix is updated

        :caseautomation: notautomated

        :caseimportance: low

### PR DESCRIPTION
**Pull request by jameerpathan111**
Fixes #5557 
Fixes #5558 
Fixes #5559 

Result:

```
$ pytest tests/foreman/cli/test_setting.py -k test_positive_update_email_reply_address 
========================================================================================================================= test session starts ==========================================================================================================================

collecting 18 items                                                                                                                                                                                                                                                   
collected 18 items / 17 deselected                                                                                                                                                                                                                                     

tests/foreman/cli/test_setting.py .                                                                                                                                                                                                                              [100%]

=============================================================================================================== 1 passed, 17 deselected in 11.55 seconds ===============================================================================================================
```

```
$ pytest tests/foreman/cli/test_setting.py -k test_negative_update_email_reply_address
========================================================================================================================= test session starts ==========================================================================================================================
collecting 18 items                                                                                                                                                                                                                                                   
collected 18 items / 17 deselected                                                                                                                                                                                                                                     

tests/foreman/cli/test_setting.py .                                                                                                                                                                                                                              [100%]

=============================================================================================================== 1 passed, 17 deselected in 6.02 seconds ================================================================================================================
```

```
$ pytest tests/foreman/cli/test_setting.py -k test_positive_update_email_subject_prefix
========================================================================================================================= test session starts ==========================================================================================================================

collecting 18 items                                                                                                                                                                                                                                                    
collected 18 items / 17 deselected                                                                                                                                                                                                                                     

tests/foreman/cli/test_setting.py .                                                                                                                                                                                                                              [100%]

=============================================================================================================== 1 passed, 17 deselected in 10.61 seconds ===============================================================================================================
```

### CODE REVIEW

File: tests/foreman/cli/test_setting.py [hunk 1]
```diff
@@ -22,8 +22,9 @@
 from robottelo.datafactory import (
     gen_string,
     generate_strings_list,
-    valid_url_list,
-    valid_data_list, invalid_boolean_strings, xdist_adapter)
+    valid_data_list,
```
Comment (oshtaier) [inline tests/foreman/cli/test_setting.py]:
```
from robottelo.datafactory import (
    gen_string,
    generate_strings_list,
    invalid_boolean_strings
    invalid_emails_list
    valid_data_list,
    valid_emails_list, 
    valid_url_list,
    xdist_adapter,
)
```
looks better

File: tests/foreman/cli/test_setting.py [hunk 2]
```diff
@@ -233,10 +246,15 @@ def test_negative_update_email_reply_address(self):
 
         :caseimportance: low
 
-        :caseautomation: notautomated
+        :caseautomation: automated
         """
+        with self.assertRaises(CLIReturnCodeError):
+            email_reply_address_value = invalid_emails_list()
```
Comment (oshtaier) [inline tests/foreman/cli/test_setting.py]: shouldn't you use similar loop construct to one you used in previous test case: `for email in invalid_emails_list():`?

File: tests/foreman/cli/test_setting.py [hunk 3]
```diff
@@ -245,10 +263,20 @@ def test_positive_update_email_subject_prefix(self):
...
         """
+        email_subject_prefix_value = gen_string('alpha')
+        Settings.set({
+            'name': "email_subject_prefix",
+            'value': email_subject_prefix_value
+        })
+        email_subject_prefix = Settings.list({
+            'search': 'name=email_subject_prefix'
+        })[0]
+        self.assertEqual(email_subject_prefix_value,
```
Comment (oshtaier) [inline tests/foreman/cli/test_setting.py]:
```
self.assertEqual(
    email_subject_prefix_value,
    email_subject_prefix['value']
)
```
looks better

File: tests/foreman/cli/test_setting.py [hunk 4]
```diff
@@ -217,10 +217,23 @@ def test_positive_update_email_reply_address(self):
...
         :caseimportance: low
 
-        :caseautomation: notautomated
+        :caseautomation: automated
         """
+        for email in valid_emails_list():
+            with self.subTest(email):
+                # The email must be escaped because some characters to not fail
+                # the parsing of the generated shell command
+                escaped_email = email.replace('"', r'\"')\
```
Comment (oshtaier) [inline tests/foreman/cli/test_setting.py]: that is not good style, probably:
```
escaped_email = email.replace(
    '"', r'\"').replace('`', r'\`')
```

Comment (jameerpathan111) [reply tests/foreman/cli/test_setting.py]: Thanks, will do that

File: tests/foreman/cli/test_setting.py (hunk 2 above)
Comment (jameerpathan111) [reply tests/foreman/cli/test_setting.py]: thanks @oshtaier , I will change it

File: tests/foreman/cli/test_setting.py (hunk 3 above)
Comment (jameerpathan111) [reply tests/foreman/cli/test_setting.py]: yes, it look better this way.
thanks @oshtaier for your comments.

File: tests/foreman/cli/test_setting.py [hunk 5]
```diff
@@ -217,10 +221,23 @@ def test_positive_update_email_reply_address(self):
...
+                # the parsing of the generated shell command
+                escaped_email = email.replace(
+                    '"', r'\"').replace('`', r'\`')
+                Settings.set({
+                    'name': "email_reply_address",
+                    'value': escaped_email})
+                email_reply_address = Settings.list({
+                    'search': 'name=email_reply_address'
+                })[0]
+                email = email.replace(r'"', '')
```
Comment (ldjebran) [inline tests/foreman/cli/test_setting.py]: why we need to replace the " char ? I was supposing that email should be saved with that char

Comment (ldjebran) [reply tests/foreman/cli/test_setting.py]:
```bash
[root@xxx]# hammer settings set  --name=email_reply_address --value \"jjjsjs\"@example.com
Setting [email_reply_address] updated to ["jjjsjs"@example.com].
```
```bash
[root@xxx]# hammer settings list --search  name=email_reply_address
--------------------|----------------------|-------------------------------------------------------
NAME                | VALUE                | DESCRIPTION                                           
--------------------|----------------------|-------------------------------------------------------
email_reply_address | "jjjsjs"@example.com | Email reply address for emails that Foreman is sending
--------------------|----------------------|-------------------------------------------------------
[root@xxx]# 
```

Comment (jameerpathan111) [reply tests/foreman/cli/test_setting.py]: thanks @ldjebran for comment.
I will again look into it. 
I am marking it as stubbed till i have a fix solution for it.

Comment (sghai) [reply tests/foreman/cli/test_setting.py]: @jameerpathan111 : Hey, please don't stub the test. If I understand correctly the issue is that when you save the value in variable the quotes automatically get removed and test failed while assertion. To get rid of failure, you used the above statement `email = email.replace(r'"', '')`

@ldjebran : do you see what's causing the quotes to be removed from email when saving to a variable ?

Comment (jameerpathan111) [reply tests/foreman/cli/test_setting.py]: I looked into it but, I was not able to get what's causing this to happen

Comment (jameerpathan111) [reply tests/foreman/cli/test_setting.py]: now i get it, basically csv output format was causing this to happen.
If we use json format it works well.
thank you @ldjebran for looking into it.
I will do the changes you have suggested.

Comment (sghai) [reply tests/foreman/cli/test_setting.py]: @ldjebran :+1:

**Comment by jameerpathan111** [comment]
Output:

```
 $  pytest tests/foreman/cli/test_setting.py -v -k "test_positive_update_email_reply_address or test_negative_update_email_reply_address or test_positive_update_email_subject_prefix"
========================================================================= test session starts ==========================================================================

collecting 18 items                                                                                                                                                   
collected 18 items / 15 deselected                                                                                                                                     

tests/foreman/cli/test_setting.py::SettingTestCase::test_negative_update_email_reply_address PASSED                                                              [ 33%]
tests/foreman/cli/test_setting.py::SettingTestCase::test_positive_update_email_reply_address PASSED                                                              [ 66%]
tests/foreman/cli/test_setting.py::SettingTestCase::test_positive_update_email_subject_prefix PASSED                                                             [100%]

=============================================================== 3 passed, 15 deselected in 31.02 seconds ===============================================================
```

**Comment by jameerpathan111** [comment]
thanks @ldjebran  for approving PR

**Comment by oshtaier** [comment]
@jameerpathan111 
good job

**Comment by jameerpathan111** [comment]
@oshtaier 
thank you

**Comment by omaciel** [comment]
Good job @jameerpathan111 

approved

**Comment by jameerpathan111** [comment]
thank you @omaciel :+1:

**Pull request merged**